src/discordtipbot/cache.cr

```diff
@@ -0,0 +1,84 @@
+class Cache
+  # Container for a snowflake and the last time it was seen
+  record Entry, id : UInt64, seen : Time do
+    def ==(other)
+      id == other.id
+    end
+
+    def <(other)
+      seen < other.seen
+    end
+
+    def >(other)
+      seen > other.seen
+    end
+
+    def hash
+      id.hash
+    end
+
+    def older_than?(age : Time::Span)
+      Time.now > seen + age
+    end
+  end
+
+  @cache = {} of UInt64 => Set(Entry)
+
+  def initialize(@ttl : Time::Span)
+  end
+
+  # Marks the `user_id` as being observed in `channel_id` at the given `time`
+  def touch(channel_id : UInt64, user_id : UInt64, time = Time.now)
+    entry = Entry.new(user_id, time)
+    set = @cache[channel_id] ||= Set(Entry).new
+    set.delete(entry) if set.includes?(entry)
+    set.add(entry)
+    @cache[channel_id] = set
+  end
+
+  # Only marks `user_id` as being observed in `channel_id` if `time` is newer then latest entry
+  def add_if_youngest(channel_id : UInt64, user_id : UInt64, time : Time)
+    entry = Entry.new(user_id, time)
+    set = @cache[channel_id] ||= Set(Entry).new
+    set.delete(entry) if set.any? { |x| x == entry && entry > x }
+    set.add(entry)
+    @cache[channel_id] = set
+  end
+
+  # Returns the set of observed entries in the given `channel_id` that have not expired.
+  def resolve(channel_id : UInt64)
+    prune(channel_id)
+  end
+
+  def resolve_to_id(channel_id : UInt64)
+    users = Set(UInt64).new
+    set = resolve(channel_id)
+    return unless set
+    set.map do |x|
+      users << x.id
+    end
+
+    users
+  end
+
+  # Clears cached sets of expired entries for give channel_id
+  def prune(channel_id : UInt64)
+    set = prune_set(@cache[channel_id])
+    @cache[channel_id] = set if set
```

Comment:
Why? I think you're missing that `@cache[channel_id]` returns a reference, not a copy.

Reply:
Yep. Where do is it documented that it's a reference?

Reply:
It's stored in a `Hash`, which is a `Class`, not a `Struct` https://carc.in/#/r/3iwc